mysql/resource_grant.go

```diff
@@ -16,16 +17,225 @@ import (
+	kTable     ObjectT = "TABLE"
+)
+
+type MySQLGrant interface {
+	GetId() string
+	SQLGrantStatement() string
+	SQLRevokeStatement() string
+	GetUserOrRole() UserOrRole
+	GrantOption() bool
+}
+
+type MySQLGrantWithDatabase interface {
+	MySQLGrant
+	GetDatabase() string
+}
+
+type MySQLGrantWithTable interface {
+	MySQLGrantWithDatabase
+	GetTable() string
+}
+
+type MySQLGrantWithPrivileges interface {
+	MySQLGrant
+	GetPrivileges() []string
+}
+
+type PrivilegesPartiallyRevocable interface {
+	SQLPartialRevokePrivilegesStatement(privilegesToRevoke []string) string
+}
+
+type UserOrRole struct {
+	Name string
+	Host string
+}
+
+func (u UserOrRole) IDString() string {
+	if u.Host == "" {
+		return u.Name
+	}
+	return fmt.Sprintf("%s@%s", u.Name, u.Host)
+}
+
+func (u UserOrRole) SQLString() string {
+	if u.Host == "" {
+		return fmt.Sprintf("'%s'", u.Name)
+	}
+	return fmt.Sprintf("'%s'@'%s'", u.Name, u.Host)
+}
+
+func (u UserOrRole) Equals(other UserOrRole) bool {
+	if u.Name != other.Name {
+		return false
+	}
+	if (u.Host == "" || u.Host == "%") && (other.Host == "" || other.Host == "%") {
+		return true
+	}
+	return u.Host == other.Host
+}
+
+type TablePrivilegeGrant struct {
 	Database   string
 	Table      string
 	Privileges []string
+	Grant      bool
+	UserOrRole UserOrRole
+	TLSOption  string
+}
+
+func (t *TablePrivilegeGrant) GetId() string {
+	return fmt.Sprintf("%s:%s", t.UserOrRole.IDString(), t.GetDatabase())
+}
+
+func (t *TablePrivilegeGrant) GetUserOrRole() UserOrRole {
+	return t.UserOrRole
+}
+
+func (t *TablePrivilegeGrant) GrantOption() bool {
+	return t.Grant
+}
+
+func (t *TablePrivilegeGrant) GetDatabase() string {
+	if strings.Compare(t.Database, "*") != 0 && !strings.HasSuffix(t.Database, "`") {
+		return fmt.Sprintf("`%s`", t.Database)
+	}
+	return t.Database
+}
+
+func (t *TablePrivilegeGrant) GetTable() string {
+	if t.Table == "*" || t.Table == "" {
+		return "*"
+	} else {
+		return fmt.Sprintf("`%s`", t.Table)
+	}
+}
+
+func (t *TablePrivilegeGrant) GetPrivileges() []string {
+	return t.Privileges
+}
+
+func (t *TablePrivilegeGrant) SQLGrantStatement() string {
+	stmtSql := fmt.Sprintf("GRANT %s ON %s.%s TO %s", strings.Join(t.Privileges, ", "), t.GetDatabase(), t.GetTable(), t.UserOrRole.SQLString())
+	if t.TLSOption != "" && strings.ToLower(t.TLSOption) != "none" {
+		stmtSql += fmt.Sprintf(" REQUIRE %s", t.TLSOption)
+	}
+	if t.Grant {
+		stmtSql += " WITH GRANT OPTION"
+	}
+	return stmtSql
+}
+
+func (t *TablePrivilegeGrant) SQLRevokeStatement() string {
+	return fmt.Sprintf("REVOKE %s ON %s.%s FROM %s", strings.Join(t.Privileges, ", "), t.GetDatabase(), t.GetTable(), t.UserOrRole.SQLString())
+}
+
+func (t *TablePrivilegeGrant) SQLPartialRevokePrivilegesStatement(privilegesToRevoke []string) string {
+	stmt := fmt.Sprintf("REVOKE %s ON %s.%s FROM %s", strings.Join(privilegesToRevoke, ", "), t.GetDatabase(), t.GetTable(), t.UserOrRole.SQLString())
+	if t.Grant {
+		stmt += " WITH GRANT OPTION"
```

Comment:
Grant option is revoked differently, like:
```
REVOKE SELECT, GRANT OPTION ON `db`.`table` FROM 'user'@'host'
```
I handled it correctly before, so this maybe (also) misses a real test, where it would be checked.

Reply:
Yeah, this is an issue.

I modified the existing test suite to catch this.